From: SeongJae Park <sj@kernel.org>
Date: Thu, 19 Feb 2026 18:08:43 -0800
Subject: [PATCH] mm/damon/core: apply min_nr_regions right after
 kdamond_merge_regions()

The upper size limit of each region is derived from min_nr_regions and
total size of monitoring regions, which can be updated while DAMON is
running.  Users could do the update in damon_call().  Underlying ops can
do that in ->update() callback.  Refresh and apply the limit right after
kdamond_merge_regions().

This is suboptimal, but DAMON is normally ok to be suboptimal by the
design.  We can optimize later if this really turns out to be a problem.

Signed-off-by: SeongJae Park <sj@kernel.org>
---
 mm/damon/core.c | 6 ++++--
 1 file changed, 4 insertions(+), 2 deletions(-)

diff --git a/mm/damon/core.c b/mm/damon/core.c
index cecddb16a313b..fb14f2507ec86 100644
--- a/mm/damon/core.c
+++ b/mm/damon/core.c
@@ -3371,10 +3371,13 @@ static int kdamond_fn(void *data)
 			max_nr_accesses = ctx->ops.check_accesses(ctx);
 
 		if (time_after_eq(ctx->passed_sample_intervals,
-					next_aggregation_sis))
+					next_aggregation_sis)) {
 			kdamond_merge_regions(ctx,
 					max_nr_accesses / 10,
 					sz_limit);
+			/* online updates might be made */
+			sz_limit = damon_apply_min_nr_regions(ctx);
+		}
 
 		/*
 		 * do kdamond_call() and kdamond_apply_schemes() after
@@ -3434,7 +3437,6 @@ static int kdamond_fn(void *data)
 				sample_interval;
 			if (ctx->ops.update)
 				ctx->ops.update(ctx);
-			sz_limit = damon_region_sz_limit(ctx);
 		}
 	}
 done:
-- 
2.47.3

